File: tests/multitrack.test.py
import os,sys,pathlib,tempfile,subprocess,copy,array,math
with tempfile.TemporaryDirectory(prefix='pve-multitrack-') as tmp:
 os.environ['PVE_DATA']=tmp;root=pathlib.Path(__file__).resolve().parent.parent;sys.path.insert(0,str(root/'engine'));import core
 def video(name,color,freq):
  path=pathlib.Path(tmp)/(name+'.mp4');subprocess.run(core.BASE+['-f','lavfi','-i',f'color=c={color}:s=160x90:r=30:d=2','-f','lavfi','-i',f'sine=frequency={freq}:sample_rate=48000:duration=2','-c:v','libx264','-pix_fmt','yuv420p','-c:a','aac','-shortest',path],capture_output=True,check=True);return core.inspect(path,name,path.name)
 def sound(name,freq):
  path=pathlib.Path(tmp)/(name+'.wav');subprocess.run(core.BASE+['-f','lavfi','-i',f'sine=frequency={freq}:sample_rate=48000:duration=1','-c:a','pcm_s16le',path],capture_output=True,check=True);return core.inspect(path,name,path.name)
 def clip(m,layer):return {'id':m['id'],'media':m['id'],'in':0,'out':2,'start':0,'layer':layer,'speed':1,'endSpeed':1,'curve':'constant','stabilization':'OFF','audio':{'volume':.4,'mute':False},'color':{}}
 media=[video('red','red',220),video('blue','blue',330),video('green','green',550)]+[sound('tone'+str(k),f) for k,f in enumerate([440,660,880,1100])]
 p={'version':1,'media':media,'clips':[clip(m,i) for i,m in enumerate(media[:3])],'audioClips':[{**clip(m,i),'kind':'audio','out':1.5,'start':i*.1,'loop':True} for i,m in enumerate(media[3:])],'audioTracks':[{} for _ in range(4)],'texts':[],'bgm':{},'aspect':'Original','export':{'resolution':'Source','fps':'30','quality':'High'}}
 p['clips'][1]['opacity']=.5;p['clips'][2]['opacity']=.5;p['clips'][0]['scaleKeyframes']=[{'time':0,'value':1},{'time':1,'value':1.35},{'time':2,'value':1}]
 def render(p,name):
  r=core.render({'id':name,'cancel':False},copy.deepcopy(p));out=core.ROOT/'exports'/r['file'];core.validate_output(out,r['duration'],True);return out,r
 def amplitudes(out):
  raw=subprocess.check_output(core.BASE+['-v','error','-ss','0.7','-i',out,'-t','0.2','-vn','-ac','1','-ar','8000','-f','f32le','pipe:1']);samples=array.array('f');samples.frombytes(raw)
  return [2*abs(sum(v*complex(math.cos(2*math.pi*f*i/8000),math.sin(2*math.pi*f*i/8000)) for i,v in enumerate(samples)))/len(samples) for f in [220,330,550,440,660,880,1100]]
 out,r=render(p,'four-tracks');rgb=list(subprocess.check_output(core.BASE+['-v','error','-ss','0.7','-i',out,'-frames:v','1','-vf','scale=1:1','-pix_fmt','rgb24','-f','rawvideo','pipe:1']));assert min(rgb)>35,rgb
 amps=amplitudes(out);assert min(amps)>.012,amps
 p['audioTracks'][2]['solo']=True;p['audioClips'][2]['out']=2.5
 out,r=render(p,'solo-tail');assert abs(r['duration']-2.7)<.04,r
 amps=amplitudes(out);assert amps[5]>.02 and max(amps[:5]+amps[6:])<.004,amps
 curve=copy.deepcopy(p);curve['clips']=[];curve['audioTracks']=[{} for _ in range(4)];curve['audioClips']=[{**clip(media[3],0),'kind':'audio','in':0,'out':1.5,'start':0,'loop':True,'audio':{'volume':.5,'gainKeyframes':[{'time':0,'value':.08},{'time':.5,'value':1.8},{'time':1.3,'value':.08}]}}];curve['bgm']={'media':media[4]['id'],'volume':.5,'fadeIn':0,'fadeOut':0,'gainKeyframes':[{'time':0,'value':.08},{'time':.5,'value':1.8},{'time':1.3,'value':.08}]};out,r=render(curve,'gain-automation');assert abs(r['duration']-1.5)<.05,r
 def tone_amp(out,at,freq=440):
  raw=subprocess.check_output(core.BASE+['-v','error','-ss',str(at),'-i',out,'-t','0.12','-vn','-ac','1','-ar','8000','-f','f32le','pipe:1']);samples=array.array('f');samples.frombytes(raw);return 2*abs(sum(v*complex(math.cos(2*math.pi*freq*i/8000),math.sin(2*math.pi*freq*i/8000)) for i,v in enumerate(samples)))/len(samples)
 low,high,low2=tone_amp(out,.12),tone_amp(out,.65),tone_amp(out,1.15);assert high>low*2 and high>low2*2,(low,high,low2);bg_low,bg_high,bg_low2=tone_amp(out,.12,660),tone_amp(out,.65,660),tone_amp(out,1.15,660);assert bg_high>bg_low*2 and bg_high>bg_low2*2,(bg_low,bg_high,bg_low2)
 print('Native three video source-audio layers, four independent tones/loops, solo, audio tails and smooth clip/BGM gain keyframes, H264/AAC full decode PASS')

File: tests/multitrack.test.mjs
import assert from 'node:assert/strict';
import {project,clip,sequence,visibleSequence,total,locate,sanitize,splitClip,audioSequence} from '../dist/model.js';
import {detachAudio,appendAudio,audioWindows,trackGain,migrateBgm} from '../dist/audio-timeline.js';
import {usage,removeUnused} from '../dist/assets.js';
const m={id:'m',duration:10,width:160,height:90,kind:'video',audio:true},p=project();p.media=[m];
p.clips=[{...clip(m),layer:0,start:0},{...clip(m),layer:1,start:1,out:2},{...clip(m),layer:2,start:2,out:1}];
assert.equal(locate(p,2.5).layer,2);assert.equal(locate(p,1.5).layer,1);assert.equal(locate(p,4).layer,0);
assert.equal(sanitize(JSON.parse(JSON.stringify(p))).clips[2].layer,2);
const source=p.clips[0],audio=detachAudio(p,source.id,0);assert(audio);assert.equal(source.audio.mute,true);assert.equal(audio.audio.mute,false);assert.equal(audio.in,source.in);
source.in=2;source.speed=2;assert.equal(audio.in,0);assert.equal(audio.speed,1);
audio.start=3;audio.out=4;const right=splitClip(audio,2);p.audioClips.push(right);assert.equal(right.start,5);assert.equal(audio.out,2);
for(let k=1;k<4;k++){const a=appendAudio(p,m,0,k);a.out=1}
assert.equal(audioWindows(p).filter(r=>r.start===0).length,3);
assert.equal(trackGain(p,3),1);p.audioTracks[2].solo=true;assert.equal(trackGain(p,3),0);assert.equal(trackGain(p,2),1);p.audioTracks[2].mute=true;assert.equal(trackGain(p,2),0);
const extra=appendAudio(p,m,12,3);extra.out=2;assert.equal(total(p),14);assert(locate(p,13).clip.gap);assert.equal(visibleSequence(p).at(-1).end,14);
const used=usage(p,m.id);assert(used>p.clips.length);p.clips=[];assert.equal(removeUnused(p,[m.id]),0);assert.equal(p.media.length,1);
const restored=sanitize(JSON.parse(JSON.stringify(p)));assert.equal(restored.audioClips.length,p.audioClips.length);assert.equal(restored.audioTracks[2].solo,true);
const legacy=project();legacy.media=[m];legacy.clips=[clip(m)];legacy.bgm={media:'m',volume:.2};migrateBgm(legacy);assert.equal(legacy.audioClips[0].out,10);assert.equal(legacy.audioClips[0].layer,2);assert(legacy.audioClips[0].loop);assert(!legacy.bgm.media);
console.log('V3 priority, independent audio split/move, four lanes/solo, audio tail, protected assets, project restore, BGM migration PASS');
